pkg/operator/csidriveroperator/driverstarter.go

```diff
@@ -0,0 +1,177 @@
+package csidriveroperator
+
+import (
+	"context"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorapi "github.com/openshift/api/operator/v1"
+	openshiftv1 "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/cluster-storage-operator/pkg/csoclients"
+	"github.com/openshift/cluster-storage-operator/pkg/generated"
+	"github.com/openshift/cluster-storage-operator/pkg/operator/csidriveroperator/csioperatorclient"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/controller/manager"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/staticresourcecontroller"
+	"github.com/openshift/library-go/pkg/operator/status"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/client-go/dynamic"
+)
+
+const (
+	conditionsPrefix = "CSIDriverStarter"
+	infraConfigName  = "cluster"
+)
+
+// This CSIDriverStarterController starts CSI driver controllers based on the underlying cloud.
+// It does not install anything by itself, only monitors Infrastructure
+// instance and starts ControllerManagers for the particular cloud.
+// It produces following Conditions:
+// CSIDriverStarterDegraded - error checking the Infrastructure
+type CSIDriverStarterController struct {
+	operatorClient v1helpers.OperatorClient
+	infraLister    openshiftv1.InfrastructureLister
+	versionGetter  status.VersionGetter
+	targetVersion  string
+	eventRecorder  events.Recorder
+
+	controllers []csiDriverControllerManager
+}
+
+type csiDriverControllerManager struct {
+	name           string
+	operatorConfig csioperatorclient.CSIOperatorConfig
+	platform       configv1.PlatformType
+	mgr            manager.ControllerManager
+	running        bool
+}
+
+func NewCSIDriverStarterController(
+	clients *csoclients.Clients,
+	dynamicClient dynamic.Interface,
+	resyncInterval time.Duration,
+	versionGetter status.VersionGetter,
+	targetVersion string,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &CSIDriverStarterController{
+		operatorClient: clients.OperatorClient,
+		infraLister:    clients.ConfigInformers.Config().V1().Infrastructures().Lister(),
+		versionGetter:  versionGetter,
+		targetVersion:  targetVersion,
+		eventRecorder:  eventRecorder,
+	}
+
+	// Populating all CSI driver operator ControllerManagers here simplifies
+	// the startup a lot
+	// - All necessary informers are populated.
+	// - CSO then just Run()s
+	//All ControllerManagers are nor running at this point! They will be started
+	c.populateCSIControllers(clients, resyncInterval, eventRecorder)
```

Comment:
IMO it'd be better to take in a `[]csiDriverControllerManager` in `NewCSIDriverStarterController` and initialize `c.controllers` here. 

It'll make it easier to test this controller, plus this package won't have any logic about which operators we're deploying.

Reply:
This is actually a good idea. I refactored the code a bit:
* CSIOperatorConfig is now a bit bigger and contains everything
* NewCSIDriverStarterController now gets array of CSIOperatorConfigs
* This array of CSIOperatorConfigs is populated in starter.go